06-classification_continued.Rmd

```diff
@@ -559,16 +657,378 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+For each set of predictors to try, we construct a model formula,
+pass it into a `recipe`, build a `workflow` that tunes
+a K-NN classifier using 5-fold cross-validation, 
+and finally records the estimated accuracy.
+
+```{r 06-fwdsel-2, warning = FALSE}
+set.seed(1)
+
+# create an empty tibble to store the results
+accuracies <- tibble(size = integer(), model_string = character(), accuracy = numeric())
+
+# create a model specification
+knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
+     set_engine("kknn") %>%
+     set_mode("classification")
+
+# create a 5-fold cross-validation object
+cancer_vfold <- vfold_cv(cancer_subset, v = 5, strata = Class)
+
+# store the total number of predictors
+n_total <- length(names)
+
+# stores selected predictors
+selected <- c()
+
+# for every size from 1 to the total number of predictors
+for (i in 1:n_total) {
+    # for every predictor still not added yet
+    accs <- list()
+    models <- list()
+    for (j in 1:length(names)) {
+        # create a model string for this combination of predictors
+        preds_new <- c(selected, names[[j]])
+        model_string <- paste("Class", "~", paste(preds_new, collapse="+"))
+
+        # create a recipe from the model string
+        cancer_recipe <- recipe(as.formula(model_string), data = cancer_subset) %>%
+                          step_scale(all_predictors()) %>%
+                          step_center(all_predictors())
+
+        # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
+        acc <- workflow() %>%
+          add_recipe(cancer_recipe) %>%
+          add_model(knn_spec) %>%
+          tune_grid(resamples = cancer_vfold, grid = 10) %>%
+          collect_metrics() %>%
+          filter(.metric == "accuracy") %>%
+          summarize(mx = max(mean))
+        acc <- acc$mx %>% unlist()
+
+        # add this result to the dataframe
+        accs[[j]] <- acc
+        models[[j]] <- model_string
+    }
+    jstar <- which.max(unlist(accs))
+    accuracies <- accuracies %>% add_row(size = i, model_string = models[[jstar]], accuracy = accs[[jstar]])
+    selected <- c(selected, names[[jstar]])
+    names <- names[-jstar]
+}
+accuracies
+```
+
+Interesting! The forward selection procedure first added the three meaningful variables `Perimeter`,
+`Concavity`, and `Smoothness`, followed by the irrelevant variables. Figure \@ref(fig:06-fwdsel-3)
+visualizes the accuracy as a function of the number of predictors in the model. You can see that
+as meaningful predictors are added, the estimated accuracy increases substantially; and as you add irrelevant
+variables, the accuracy either exhibits small fluctuations or decreases as the model attempts to tune the number
+of neighbours to account for the extra noise. In order to pick the right model from the sequence, you have 
+to trade off the desiderata to have a high accuracy, but also a simple model with few predictors (which has 
```

Comment:
desiderata is not a commonly used word and provides unnecessary cognitive load in this sentence.
```suggestion
to balance high accuracy and model simplicity - one which has a few predictors (which has 
```

Reply:
done